File: lang/lua/astal/file.lua
local lgi = require("lgi")
local Astal = lgi.require("AstalIO", "0.1")
local GObject = lgi.require("GObject", "2.0")

local M = {}

---@param path string
---@return string
function M.read_file(path)
    return Astal.read_file(path)
end

---@param path string
---@param callback fun(content: string, err: string): nil
function M.read_file_async(path, callback)
    Astal.read_file_async(path, function(_, res)
        local content, err = Astal.read_file_finish(res)
        callback(content, err)
    end)
end

---@param path string
---@param content string
---@param follow_symlinks boolean
function M.write_file(path, content, follow_symlinks)
    Astal.write_file(path, content, follow_symlinks or true)
end

---@param path string
---@param content string
---@param callback? fun(err: string): nil
---@param follow_symlinks boolean
function M.write_file_async(path, content, callback, follow_symlinks)
    Astal.write_file_async(path, content, function(_, res)
        if type(callback) == "function" then
            callback(Astal.write_file_finish(res))
        end
    end, follow_symlinks or true)
end

---@param path string
---@param callback fun(file: string, event: integer): nil
---@param follow_symlinks boolean
function M.monitor_file(path, callback, follow_symlinks)
    return Astal.monitor_file(path, GObject.Closure(callback), follow_symlinks or true)
end

return M

File: lang/python/astal/file.py
from typing import Callable

from gi.repository import AstalIO

def read_file(fp: str) -> str:
    return AstalIO.read_file(fp)

def read_file_async(fp: str, callback: Callable[[str | None, Exception | None], None]) -> None:
    try:
        AstalIO.read_file_async(fp, lambda _, res: callback(AstalIO.read_file_finish(res), None))

    except Exception as e:
        callback(None, e)

def write_file(fp: str, content: str) -> None:
    AstalIO.write_file(fp, content)

def write_file_async(fp: str, content: str, callback: Callable[[Exception], None]) -> None:
    try:
        AstalIO.write_file_async(fp, content, lambda _, res: AstalIO.write_file_finish(res))

    except Exception as e:
        callback(e)

def monitor_file(fp: str, callback: Callable[[str, int], None]) -> None:
    return AstalIO.monitor_file(fp, callback)
